pkg/apis/v1/nodepool.go

```diff
@@ -160,6 +160,77 @@ func (l Limits) ExceededBy(resources v1.ResourceList) error {
+	// +kubebuilder:validation:MaxItems:=100
+	// +required
+	Requirements []NodeSelectorRequirementWithMinValues `json:"requirements" hash:"ignore"`
+	// NodeClassRef is a reference to an object that defines provider specific configuration
+	// +required
+	NodeClassRef *NodeClassReference `json:"nodeClassRef"`
+	// TerminationGracePeriod is the maximum duration the controller will wait before forcefully deleting the pods on a node, measured from when deletion is first initiated.
+	//
+	// Warning: this feature takes precedence over a Pod's terminationGracePeriodSeconds value, and bypasses any blocked PDBs or the karpenter.sh/do-not-disrupt annotation.
+	//
+	// This field is intended to be used by cluster administrators to enforce that nodes can be cycled within a given time period.
+	// When set, drifted nodes will begin draining even if there are pods blocking eviction. Draining will respect PDBs and the do-not-disrupt annotation until the TGP is reached.
+	//
+	// Karpenter will preemptively delete pods so their terminationGracePeriodSeconds align with the node's terminationGracePeriod.
+	// If a pod would be terminated without being granted its full terminationGracePeriodSeconds prior to the node timeout,
+	// that pod will be deleted at T = node timeout - pod terminationGracePeriodSeconds.
+	//
+	// The feature can also be used to allow maximum time limits for long-running jobs which can delay node termination with preStop hooks.
+	// If left undefined, the controller will wait indefinitely for pods to be drained.
+	// +kubebuilder:validation:Pattern=`^([0-9]+(s|m|h))+$`
+	// +kubebuilder:validation:Type="string"
+	// +optional
+	TerminationGracePeriod *metav1.Duration `json:"terminationGracePeriod,omitempty"`
+	// ExpireAfter is the duration the controller will wait
+	// before terminating a node, measured from when the node is created. This
+	// is useful to implement features like eventually consistent node upgrade,
+	// memory leak protection, and disruption testing.
+	// +kubebuilder:default:="720h"
+	// +kubebuilder:validation:Pattern=`^(([0-9]+(s|m|h))+)|(Never)$`
+	// +kubebuilder:validation:Type="string"
+	// +kubebuilder:validation:Schemaless
+	// +optional
+	ExpireAfter NillableDuration `json:"expireAfter,omitempty"`
+}
+
+// This is used to convert between the NodeClaim's NodeClaimSpec to the Nodepool NodeClaimTemplate's NodeClaimSpec.
+func (in *NodeClaimTemplateSpec) ToNodeClaimSpec() *NodeClaimSpec {
+	return &NodeClaimSpec{
+		Taints:                 in.Taints,
+		StartupTaints:          in.StartupTaints,
+		Requirements:           in.Requirements,
+		NodeClassRef:           in.NodeClassRef,
+		TerminationGracePeriod: in.TerminationGracePeriod,
+		ExpireAfter:            in.ExpireAfter,
+	}
+}
+
+// NodeClaimTemplateWithResources is used in scheduling simulations
+type NodeClaimTemplateWithResources struct {
```

Comment:
Do we need this if we are just going to convert to a NodeClaim that can track resources?

Reply:
Why not just store the Resources in the scheduling.go file? Leaking the scheduling implementation into the APIs package feels wrong to me